### PR TITLE
Refactor pulling

### PR DESCRIPTION
**Pull request by calmh**
## Current progress

- [x] Files
- [x] Symlinks
- [x] Directories
- [x] Permissions
- [x] Modification times
- [x] Forcing rescan (interface)
- [ ] Forcing rescan (actual implementation)
- [x] Versioning
- [x] Virtual mtimes (interface)
- [x] Virtual mtimes (actual implementation)
- [x] Progress updates
- [x] Database updates (via Progresser interface)
- [x] Network pulling parallellization
- [x] Conflict handling
- [ ] Tricky test cases!
- [x] Integrate with rwfolder
- [ ] Benchmark to ensure we don't regress too badly
- [x] Proper queue status / progress in GUI
- [x] Manual queue reordering

## Issues

- We need the entire set of files to change in memory. We need this in order to properly sort stuff, but it may require large amounts of memory. Perhaps we can get away with doing it in segments anyway, and retrying failures...

- Manual reordering of the pull queue is no longer allowed. The reason is that we enforce an ordering on the entire change set so that we can properly do things like reuse blocks while renaming, make sure to delete a directory that is in the way for a file, and so on. However we retain as much as possible of the original ordering so newest-first, smallest-first etc still work mostly as you'd expect.

### CODE REVIEW

File: lib/fsops/changeset.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	CurrentFile(name string) (protocol.FileInfo, bool)
+}
+
+type IgnoreMatcher interface {
+	// Match returns true if the file should be ignored.
+	Match(filename string) bool
+}
+
+type Source interface {
+	Request(file string, offset int64, hash []byte, buf []byte) error
```
Comment (AudriusButkevicius) [inline lib/fsops/changeset.go]: Needs flags opts, as I plan to use that.

File: lib/fsops/changeset.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+		} else {
+			c.updateDirs = append(c.updateDirs, f)
+		}
+	} else {
+		if f.IsDeleted() {
+			key := string(f.Hash())
+			f.Blocks = nil
+			c.deleteFiles = append(c.deleteFiles, f)
+			if c.deletedHashes == nil {
+				c.deletedHashes = make(map[string]protocol.FileInfo)
```
Comment (AudriusButkevicius) [inline lib/fsops/changeset.go]: Doing this in the constructor saves us from checking this if all the time.

File: lib/fsops/changeset.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			c.updateFiles = append(c.updateFiles, f)
+		}
+	}
+}
+
+func (c *Changeset) Apply() error {
+	for _, f := range c.updateFiles {
+		var err error
+		if old, ok := c.deletedHashes[string(f.Hash())]; ok {
+			// There exists an identical file that we are supposed to delete. Lets instead rename it.
```
Comment (AudriusButkevicius) [inline lib/fsops/changeset.go]: Given the overall hash matches, I'd check the rest of the hashes one by one.

File: lib/fsops/changeset.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+			err = c.renameFile(old, f)
```
Comment (AudriusButkevicius) [inline lib/fsops/changeset.go]: Need to delete from deletedFiles?

**Comment by AudriusButkevicius** [comment]
:+1: so far.

**Comment by AudriusButkevicius** [comment]
> Manual reordering of the pull queue is no longer allowed

We should still support that, but only for files?

**Comment by calmh** [comment]
Is there a good use case for this? Has anyone ever used it in practice, now that we have the sort ordering config available?

**Comment by AudriusButkevicius** [comment]
Well I didn't come up with the feature myself, it was requested on the tracker. Your argument about fairly fast only applies to many dmall files. Picking which VM image to pull first is perfectly viable.

Is there a lot of complexity involved in this? I though the graph is only used to work out what needs doing before we can pull the file? I don't think asking to pull a specific file is essentially no extra complexity?

**Comment by calmh** [comment]
As it works right now (in this change), we schedule all the needed items, sort them based on the selected sort criteria, then reorder based on dependencies. Changing the order manually after really also requires redoing the dependency sort, which can change the order back. For example moving a file to the top may require creating the directory it's supposed to be in, which in turn may require renaming a file that's in the way for that directory, etc recursively. All this while we're already working of the queue that we're reordering, and we need the GUI to be able to reach all the way into the changeset which is a few levels away. It's certainly not impossible, but I'm unsure if it's worth it.

For sure *someone* wanted the feature, but ... we didn't have the preset "newest first", "largest first" etc options at that time either I think.

File: lib/changeset/changeset.go [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+
+func (c *Changeset) Size() int {
+	return len(c.queue)
+}
+
+// progressAccount calls fn(f) and returns it's return value, while ensuring
+// that the Progresser is called appropriately before and after.
+func (c *Changeset) progressAccount(fn func(protocol.FileInfo) *opError, f fileInfo) *opError {
+	if !f.synthetic && c.Progresser != nil {
+		c.Progresser.Started(f.FileInfo)
```
Comment (AudriusButkevicius) [inline lib/changeset/changeset.go]: You could replace 3 synthetic if branches with one early one.

**Comment by AudriusButkevicius** [comment]
So I've read the code on a small screen, and I don't understand why it's an issue apart from code which has to reach into the queue in a changeset. I understand that moving a file to the top will break the chosen sort order, especially it has deps to satisfy, but I think this is expected, and nobody would complain about that?

I guess lets get this in without that feature, but I do feel that we'll have requests for it to comeback.

File: lib/changeset/changeset.go (hunk 5 above)
Comment (calmh) [inline lib/changeset/changeset.go]: Yeah, fixed.

**Comment by calmh** [comment]
I think you're right. I have some stuff to explore here anyway, will look into it again and maybe not worry too much about the consequences as it's what the user asked for and we'll retry anyway when we failed (and we need to keep the retry logic, because there are still situations when we'll fail which I'll document some time...).

This now passes the normal integration tests, i.e. "works" for some values of works. :)

**Comment by calmh** [comment]
https://reviews.syncthing.net/cru/CR-3

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
@st-jenkins retest this please, once more with feeling!

**Comment by calmh** [comment]
@st-jenkins retest this please asdkjaslkdj asldkja ldkasj dalskdj asd

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by calmh** [comment]
@st-jenkins retest this please

**Comment by lkwg82** [comment]
I just checked out and  checked the UI. You need to remove the "file pull order" setting from the advanced folder dialog. If this is meant by "dissallow manual reorder".

**Comment by calmh** [comment]
I'll split this up and post some smaller PRs

**Pull request closed without merging**